sdks/java/core/src/main/java/org/apache/beam/sdk/io/BigQueryIO.java

```diff
@@ -571,12 +657,535 @@ public boolean getValidate() {
+        String jobId, TableReference table, JobService jobService)
+            throws InterruptedException, IOException {
+      String destinationUri = String.format("%s/%s", extractDestinationDir, "*.avro");
+      JobConfigurationExtract extract = new JobConfigurationExtract();
+      extract.setSourceTable(table);
+      extract.setDestinationFormat("AVRO");
+      extract.setDestinationUris(ImmutableList.of(destinationUri));
+
+      LOG.info("Starting BigQuery extract job: {}", jobId);
+      jobService.startExtractJob(jobId, extract);
+      Job job =
+          jobService.pollJob(table.getProjectId(), jobId, JOB_POLL_MAX_RETRIES);
+      if (parseStatus(job) != Status.SUCCEEDED) {
+        throw new IOException(String.format(
+            "Extract job %s failed, status: %s",
+            job.getJobReference().getJobId(), job.getStatus()));
+      }
+
+      JobStatistics jobStats = job.getStatistics();
+      List<Long> counts = jobStats.getExtract().getDestinationUriFileCounts();
+      if (counts.size() != 1) {
+        String errorMessage = (counts.size() == 0 ?
+            "None destination uri file count received." :
+            String.format("More than one destination uri file count received. First two are %s, %s",
+                counts.get(0), counts.get(1)));
+        throw new RuntimeException(errorMessage);
+      }
+
+      long filesCount = counts.get(0);
+      List<String> tempFiles = Lists.newArrayList();
+      IOChannelFactory factory = IOChannelUtils.getFactory(extractDestinationDir);
+      for (int i = 0; i < filesCount; ++i) {
+        String filePath =
+            factory.resolve(extractDestinationDir, String.format("%012d%s", i, ".avro"));
+        tempFiles.add(filePath);
+      }
+      return ImmutableList.copyOf(tempFiles);
+    }
+
+    private List<BoundedSource<TableRow>> createSources(
+        List<String> files, TableSchema tableSchema) throws IOException, InterruptedException {
+      final String jsonSchema = JSON_FACTORY.toString(tableSchema);
+
+      SerializableFunction<GenericRecord, TableRow> function =
+          new SerializableFunction<GenericRecord, TableRow>() {
+            @Override
+            public TableRow apply(GenericRecord input) {
+              try {
+                return AvroUtils.convertGenericRecordToTableRow(
+                    input, JSON_FACTORY.fromString(jsonSchema, TableSchema.class));
+              } catch (IOException e) {
+                throw new RuntimeException("Failed to convert GenericRecord to TableRow", e);
+              }
+            }};
+
+      List<BoundedSource<TableRow>> avroSources = Lists.newArrayList();
+      BackOff backoff = new AttemptBoundedExponentialBackOff(
+          MAX_FILES_VERIFY_ATTEMPTS, INITIAL_FILES_VERIFY_BACKOFF_MILLIS);
+      for (String fileName : files) {
+        while (BackOffUtils.next(Sleeper.DEFAULT, backoff)) {
+          if (IOChannelUtils.getFactory(fileName).getSizeBytes(fileName) != -1) {
```

Comment:
I will start another PR to add batch api in GcsUtil, so this PR won't get too large.